File: ts/add_binary.ts
function addBinary(a: string, b: string): string {
  const maxLength = Math.max(a.length, b.length);
  a = a.padStart(maxLength, '0');
  b = b.padStart(maxLength, '0');

  let result = '';
  let carry = 0;

  for (let i = maxLength - 1; i >= 0; i--) {
    const sum = Number(a[i]) + Number(b[i]) + carry;
    result = (sum % 2) + result;
    carry = Math.floor(sum / 2);
  }

  return result = carry ? carry + result : result;
}

File: easy/ruby/add_binary.rb
# @param {String} a
# @param {String} b
# @return {String}
def add_binary(a, b)
  max_len = [a.length, b.length].max
  a = a.rjust(max_len, '0')
  b = b.rjust(max_len, '0')

  res = ''
  carry = 0

  (max_len - 1).downto(0) do |i|
    sum = a[i].to_i + b[i].to_i + carry
    res = (sum % 2).to_s + res
    carry = sum / 2
  end

  carry.positive? ? carry.to_s + res : res
end
